src/com/rockbite/bootcamp/util/command/UndoBuyCommand.java

```diff
@@ -0,0 +1,42 @@
+package com.rockbite.bootcamp.util.command;
+
+import com.rockbite.bootcamp.Player;
+import com.rockbite.bootcamp.Product;
+import com.rockbite.bootcamp.ShopImpl;
+import com.rockbite.bootcamp.util.pool.Poolable;
+
+/**
+ * Command for undo buying a product from the shop
+ */
+public class UndoBuyCommand implements Command, Poolable {
```

Comment:
This command should have had kept the product ID, remembered which product was bought. And by itself would have been the history keeper of itself. So shop does not have to keep track of sold products or deep copy them.

Just needed to keep one Integer as a field for both commands you have. and create new command for each history case without freeing them from pool